crates/bevy_ecs/src/component.rs

```diff
@@ -1959,13 +2011,54 @@ impl Components {
         self.components.get(id.0).and_then(|info| info.as_ref())
     }
 
-    /// Returns the name associated with the given component, if it is registered.
-    /// This will return `None` if the id is not regiserted or is queued.
+    /// Gets the [`ComponentDescriptor`] of the component with this [`ComponentId`] if it is present.
+    /// This will return `None` only if the id is neither regisered nor queued to be registered.
+    ///
+    /// Currently, the [`Cow`] will be [`Cow::Owned`] if and only if the component is queued. It will be [`Cow::Borrowed`] otherwise.
     ///
     /// This will return an incorrect result if `id` did not come from the same world as `self`. It may return `None` or a garbage value.
     #[inline]
-    pub fn get_name(&self, id: ComponentId) -> Option<&str> {
-        self.get_info(id).map(ComponentInfo::name)
+    pub fn get_descriptor<'a>(
+        &'a self,
+        id: ComponentId,
+    ) -> Option<CowArc<'a, ComponentDescriptor>> {
+        self.components
+            .get(id.0)
+            .and_then(|info| info.as_ref().map(|info| CowArc::Borrowed(&info.descriptor)))
+            .or_else(|| {
+                let queued = self.queued.read().unwrap_or_else(PoisonError::into_inner);
+                // first check components
+                queued
+                    .components
+                    .values()
+                    .find(|queued| queued.id == id)
+                    .map(|queued| CowArc::Owned(Arc::new(queued.descriptor.clone())))
+                    .or_else(|| {
+                        // otherwise check resources
+                        queued
+                            .resources
+                            .values()
+                            .find(|queued| queued.id == id)
+                            .map(|queued| CowArc::Owned(Arc::new(queued.descriptor.clone())))
+                    })
+                    .or_else(|| {
+                        // otherwise check dynamic
+                        queued
+                            .dynamic_registrations
+                            .iter()
+                            .find(|queued| queued.id == id)
+                            .map(|queued| CowArc::Owned(Arc::new(queued.descriptor.clone())))
+                    })
```

Comment:
Since you do the same thing with each `Iterator`, it might be more clear to combine them with `chain()` and do `find()`/`map()` once: 

```suggestion
                // first check components, then resources, then dynamic
                queued
                    .components
                    .values()
                    .chain(queued.resources.values())
                    .chain(queued.dynamic_registrations.iter())
                    .find(|queued| queued.id == id)
                    .map(|queued| CowArc::Owned(Arc::new(queued.descriptor.clone())))
```